pylint/checkers/unicode.py

```diff
@@ -0,0 +1,701 @@
+            # If everything is correctly decoded, the BOM will be removed automatically
+            return _map_positions_to_result(line_search, search_dict, "\n")
+        except UnicodeDecodeError:
+            # If we can't decode properly, we simply use bytes, even so the column offsets
+            # might be wrong a bit, but it is still better then nothing
+            line_search_byte = line
+            search_dict_byte = {
+                _cached_encode_search(char.unescaped, codec): char
+                for char in cls._BAD_CHARS
+            }
+            # Using bytes we have to remove the BOM manually
+            return _map_positions_to_result(
+                line_search_byte,
+                search_dict_byte,
+                _cached_encode_search("\n", codec),
+                byte_str_length=_byte_to_str_length(codec),
+            )
+
+    def process_module(self, node: nodes.Module) -> None:
+        """Perform the actual check by checking module stream
+        the module's content is accessible via node.stream() function
+        """
+
+        # Check for UTF-16/UTF-32, to warn for it
+        lineno: int
+        line: bytes
+        codec: Optional[str]
+        lines: Sequence[bytes]
+        invalid_codec_line: Optional[int] = None
+
+        # First try to detect encoding with PEP 263, won't work
+        # see https://bugs.python.org/issue1503789
+        # pylint: disable-next=fixme
+        # TODO we could manually modify the stream to convert UTF16/UTF32 to ASCII
+        #      but why ;-) ?
+        try:
+            with node.stream() as stream:
+                codec, lines = detect_encoding(stream.readline)
+        except SyntaxError:
+            pass
+        else:
+            if self._is_invalid_codec(codec):
+                invalid_codec_line = len(lines)
+
+        # If there is no invalid codec already, lets try to determine a BOM,
+        # even so Python currently doesn't like non UTF-8 BOMs ....
+        if invalid_codec_line is None:
+            with node.stream() as stream:
+                line = stream.readline()
+                codec = extract_codec_from_bom(line)
+            if self._is_invalid_codec(codec):
+                invalid_codec_line = 1
+
+        if invalid_codec_line is not None:
+            self.add_message(
+                "invalid-unicode-codec",
+                # Currently Nodes will lead to crashes of pylint
+                # node=node,
+                line=invalid_codec_line,
+                end_lineno=invalid_codec_line,
+                confidence=interfaces.HIGH,
+                col_offset=None,
+                end_col_offset=None,
+            )
+
+        # Default to UTF-8 if no codec was found
+        codec = codec or "UTF-8"
+
+        # Check for Bidirectional Unicode
+        with node.stream() as stream:
+            for (lineno, line) in enumerate(_fix_utf16_32_line_stream(stream, codec)):
```

Comment:
If we make this a method we can do it in the previous `node.stream()` call and don't have to open the stream twice.

Reply:
Sorry I don't get this.
How would making this a method prevent us from requiring  to open the stream again?
You mean that we could do only a single node.stream() and use seek before all calls that expect to start from the beginning?

Reply:
Ok did it.